### ISSUE TITLE
Integration with google/oss-fuzz

### ISSUE THREAD

**Issue by silvergasp** [Byron/gitoxide#703]
### Summary 💡

First of all thanks for all your excellent work here, I'm a huge fan of your work :) Google's [OSS-fuzz](https://github.com/google/oss-fuzz) is a free continuous fuzzing service. If you are interested in integrating this gitoxide with OSS-fuzz, the OSS-fuzz project will do the following each day;
- Build each of your fuzz tests,
- Run each of your fuzz tests on a distributed cluster,
- Report bugs back to you via email or via the private bug tracker.

I'd be happy to champion the effort for integration if you are keen on it?

### Motivation 🔦

I'd like to help with improving the stability of gix, by running all the fuzz tests using googles free cluster/fuzzing platform :)

**Comment by Byron** [comment]
That sounds fantastic, please go ahead. There are already some fuzz setups, mostly for parsers, that you might be able to (re-)use. Adding more might also be a good idea, there are several parsers without a fuzzer setup. `git-index` comes to mind as a more complex one, and certainly the ones for parsing objects in `git-object` as well.

> Report bugs back to you via email or via the private bug tracker.

Feel free to report them here on GitHub issues if that's an option, otherwise my email address will work as well.

**Comment by silvergasp** [comment]
> That sounds fantastic, please go ahead. There are already some fuzz setups, mostly for parsers, that you might be able to (re-)use. Adding more might also be a good idea, there are several parsers without a fuzzer setup. `git-index` comes to mind as a more complex one, and certainly the ones for parsing objects in `git-object` as well.
> 
> > Report bugs back to you via email or via the private bug tracker.
> 
> Feel free to report them here on GitHub issues if that's an option, otherwise my email address will work as well.

Not a problem. I'll configure OSS-fuzz to report to the GitHub issue tracker. The only downside to configuring OSS-fuzz in this way is that **if** a vulnerability is found in gitoxide, then it will be publicly reported. Feel free to let me know if you change your mind and I'll tweak the oss-fuzz integration.

A couple of things I'll need from you;
- An email address: [This should preferably be one associated with a google account](https://google.github.io/oss-fuzz/faq/#why-do-you-require-a-google-account-for-authentication). Your `@icloud.com` email address in the git logs will work for receiving bug reports but you won't be able to get full access to the fuzzing statistics dashboard or the private bug tracker. Also, something to keep in mind is that the email address you choose will be stored in plain text in the oss-fuzz repository.
- I'll need you to comment on the PR (I'll post a link here once I'm done), mentioning that you approve/dissaprove of the integration.

**Comment by Byron** [comment]
> Feel free to let me know if you change your mind and I'll tweak the oss-fuzz integration.

I dare it :) - and if a bug is discovered I will be quick with a patch and patch-release.

> An email address: [This should preferably be one associated with a google account](https://google.github.io/oss-fuzz/faq/#why-do-you-require-a-google-account-for-authentication).

`cat .mailmap` in the root of this repository reveals my Google email address.

> I'll need you to comment on the PR (I'll post a link here once I'm done), mentioning that you approve/dissaprove of the integration.

Alright, I am looking forward to it! Thanks again.

**Comment by silvergasp** [comment]
> > Feel free to let me know if you change your mind and I'll tweak the oss-fuzz integration.
> 
> I dare it :) - and if a bug is discovered I will be quick with a patch and patch-release.
> 
> > An email address: [This should preferably be one associated with a google account](https://google.github.io/oss-fuzz/faq/#why-do-you-require-a-google-account-for-authentication).
> 
> `cat .mailmap` in the root of this repository reveals my Google email address.
> 
> > I'll need you to comment on the PR (I'll post a link here once I'm done), mentioning that you approve/dissaprove of the integration.
> 
> Alright, I am looking forward to it! Thanks again.

Easy, the PR is here https://github.com/google/oss-fuzz/pull/9529

**Comment by silvergasp** [comment]
Thank you for being so quick to respond with issues/PR's. The OSS-fuzz team may take a few days to review the PR but I'm fairly confident that it will get merged . After which point you should be able to get access to fuzzing statistics and crash reports at https://oss-fuzz.com/

If/when all that goes through successfully I'll open a PR here to add a GitHub action which will trigger fuzzing on every PR :)

**Comment by Byron** [comment]
> If/when all that goes through successfully I'll open a PR here to add a GitHub action which will trigger fuzzing on every PR :)

I wonder if this can be efficient. Maybe it counts as efficient as they will spend X compute time without repeating previous input, or while continuing from a previous fuzzer state as much as possible. Otherwise it would seem wasteful if fuzzing occours even though nothing might have changed. I am saying this because most of the fuzzed crates don't change much (at least right now).

**Comment by silvergasp** [comment]
Yeah, oss-fuzz is configured to "save its state" to a fuzz corpus so that the fuzzer can continue where it's left off. So I wouldn't say it's going to be inefficient to get set up with ci-fuzz. There are two components to this;
- OSS-fuzz which runs for X compute time per day (Typically hours across a distributed cluster).
- CIFuzz which runs for Y compute time per PR (Configurable, but set to 10min by default)

CIFuzz is configured to download the fuzz corpus from OSS-fuzz, so it has the benefit of the entire fuzz corpus generated from millions of inputs. The goal of CIFuzz is to catch "shallow" bugs, before they get into the gitoxide repository. Whereas the goal of OSS-fuzz is to catch deeper bugs, that wouldn't otherwise be found in 10min of fuzzing.

It's worth noting that the sharing of the state is typically configured in one direction oss-fuzz->ci-fuzz. This works out well for most cases, as oss-fuzz will explore much deeper into your code than ci-fuzz will.

**Comment by Byron** [comment]
Thanks for elaborating, this sounds quite sane to me and makes me feel I don't unnecessarily heat the planet 😅.

**Comment by silvergasp** [comment]
Yeah that's fair :) If it helps, [github has been carbon neutral since 2019](https://github.blog/2021-04-22-environmental-sustainability-github/) and google cloud (where oss-fuzz runs) [is also carbon nuetral](https://cloud.google.com/sustainability/).

So the approximate "carbon diff" for this set of changes;
- OSS-fuzz (google cloud): net 0 
- CI-fuzz (github actions): net 0

**Comment by silvergasp** [comment]
On a different note, the initial integration is now complete and you should begin seeing results pop up on https://oss-fuzz.com/ in the next 24hrs.

**Comment by Byron** [comment]
Thanks for the heads-up! I probably have to check how it's possible to be carbon neutral but not use 100% renewable energy (in GitHub's case), but that's unrelated to how Google seems to be completely green which seems nice.

> On a different note, the initial integration is now complete and you should begin seeing results pop up on https://oss-fuzz.com/ in the next 24hrs.

I tried to login with my google account but it put me straight back to the login screen without any message (twice). The same happened with the GitHub integration.

**Comment by silvergasp** [comment]
It does look like the integration has worked successfully https://oss-fuzz-build-logs.storage.googleapis.com/index.html#gitoxide so fuzzing results should be available soon.

Can you confirm that this is correct? https://github.com/google/oss-fuzz/blob/master/projects/gitoxide/project.yaml#L3

**Comment by Byron** [comment]
> If I end up adding more fuzz harnesses to gitoxide I may request that I get access to the dashboard as well. It's as simple as modifying the project.yaml in the oss-fuzz project to include my email address as well.

Thanks, I will give it some time then and escalate if needed.

> Can you confirm that this is correct? https://github.com/google/oss-fuzz/blob/master/projects/gitoxide/project.yaml#L3

Yes, it's correct and it already found some bugs :). It created issues and sent me an email, so all good. Technically I wouldn't need emails, but definitely want to have issues, let's see how this works out in the end.

**Comment by silvergasp** [comment]
> > If I end up adding more fuzz harnesses to gitoxide I may request that I get access to the dashboard as well. It's as simple as modifying the project.yaml in the oss-fuzz project to include my email address as well.
> 
> Thanks, I will give it some time then and escalate if needed.
> 
> > Can you confirm that this is correct? https://github.com/google/oss-fuzz/blob/master/projects/gitoxide/project.yaml#L3
> 
> Yes, it's correct and it already found some bugs :). It created issues and sent me an email, so all good. Technically I wouldn't need emails, but definitely want to have issues, let's see how this works out in the end.

Oh, nice! Well, it's encouraging to see that it was worth the effort.

In the recent CI-fuzz PR I made, I forgot to mention that there are some more detailed docs for that github action here https://google.github.io/oss-fuzz/getting-started/continuous-integration/.

**Comment by silvergasp** [comment]
Are you able to get access to the private chromium bug tracker?
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=55559
You should be able to download the failing test case from there if you can't yet login to the dashboard.

**Comment by Byron** [comment]
Yes, that works. The oss-fuzz.com page still doesn't, it looks like the underpinnings are different. I will wait, there is nothing stopping me from fixing these bugs I think if the reproductions work.

**Comment by silvergasp** [comment]
Just to continue on from the conversations here https://github.com/Byron/gitoxide/pull/714#issuecomment-1411580869. Is there areas where you'd want to prioritise adding fuzz harnesses?

Are there crates where it just doesn't make sense to add fuzzing at all?

Now that all the current fuzz tests are working do you want new fuzzers to be automatically added to oss fuzz via a glob?

**Comment by Byron** [comment]
Thanks for bringing it (back) up!

> Is there areas where you'd want to prioritise adding fuzz harnesses?

Generally everything that `gitoxide` has to read from disk and where a parser is implemented by hand, i.e. without `nom`. This would mean the following crates could use fuzzing:

* `git-index` for reading indices (i.e. `.git/index`)
* `git-ref` for reading packed refs (i.e. `.git/packed-refs`)
* `git-pack` for reading `*.pack`, `*.idx`, and multi-pack indices.

That should be about it.

> Are there crates where it just doesn't make sense to add fuzzing at all?

I don't know actually. Working with include-lists is still easier here.

> Now that all the current fuzz tests are working do you want new fuzzers to be automatically added to oss fuzz via a glob?

I think that makes perfect sense, I'd appreciate this. Thank you!

**Comment by silvergasp** [comment]
> Generally everything that `gitoxide` has to read from disk and where a parser is implemented by hand, i.e. without `nom`. This would mean the following crates could use fuzzing:
> 
> * `git-index` for reading indices (i.e. `.git/index`)
> * `git-ref` for reading packed refs (i.e. `.git/packed-refs`)
> * `git-pack` for reading `*.pack`, `*.idx`, and multi-pack indices.

Easy I think I'll have a bit of a go at tackling git-pack, as that currently doesn't have any fuzz tests.

> > Now that all the current fuzz tests are working do you want new fuzzers to be automatically added to oss fuzz via a glob?
> 
> I think that makes perfect sense, I'd appreciate this. Thank you!

Not a problem I'ved opened a PR here https://github.com/google/oss-fuzz/pull/9553

**Comment by silvergasp** [comment]
Hey mate, I'm going to go through and start triaging some of the oss-fuzz bugs. But I thought I'd connect with you first regarding your goals/non-goals for the project. I've read through "Project Goals" in the README.md, and I think the relevant part to OSS-fuzz integration is this;
> Take shortcuts, but not in quality
> - binaries may use anyhow::Error exhaustively, knowing these errors are solely user-facing.
> - libraries use light-weight custom errors implemented using quick-error or thiserror.
> - internationalization is nothing we are concerned with right now.
> - IO errors due to insufficient amount of open file handles don't always lead to operation failure

Beyond this do you have any guidance on crates that I should prioritise when triaging, and what kinds of bugs should take precedence over others? It's ok if you don't have an immediate answer/are still formulating one. As a starting point here is my best guess for priorities (lower number == higher priority);

1. Security vulnerabilities in any crate,
2. Bugs in ["production grade" code](https://github.com/Byron/gitoxide#production-grade),
3. Bugs in ["Stabilisation candidates"](https://github.com/Byron/gitoxide#stabilization-candidates),
4. Bugs in ["Unstable/initial development"](https://github.com/Byron/gitoxide#initial-development),
5. Bugs in third-party dependencies.

Let me know if you think that I'm off base or if you'd like to prioritise specific areas of code :)

**Comment by Byron** [comment]
> Hey mate, I'm going to go through and start triaging some of the oss-fuzz bugs.

Great, thanks so much! The last `git-revspec::parse` stackoverflow error I couldn't reproduce, strangely enough, I think I commented this [here](https://github.com/Byron/gitoxide/issues/711#issuecomment-1414334797).

Your prioritization seems sound, and I think you should do what makes sense to you first and foremost, I'd probably rather follow your guidance.

Let me just share one thought I had a couple of times: `gitoxide` depends heavily on the `time` crate, a crate which I think is heavily used in the ecosystem but definitely not as heavily fuzzed. When having to [handle panics](https://github.com/Byron/gitoxide/commit/786f6dc5c1f765b9598cd55ca8fb1714ad177e46#diff-be98a99b296e2c61bcf1c19878a39bbee95e035af7b0fb3f56805de9a0dc1fe9R120) in the time crate I don't feel very comfortable and think that the `time` crate probably deserves more attention by fuzzers than it has now. Or in other words, I wouldn't want to fuzz `time` by needling the fuzzer through the API of `git_date::parse()` :D. Maybe there are more direct ways to improve it.

That's all, happy fuzzing!

**Comment by silvergasp** [comment]
>Let me just share one thought I had a couple of times: gitoxide depends heavily on the time crate, a crate which I think is heavily used in the ecosystem but definitely not as heavily fuzzed. When having to [handle panics](https://github.com/Byron/gitoxide/commit/786f6dc5c1f765b9598cd55ca8fb1714ad177e46#diff-be98a99b296e2c61bcf1c19878a39bbee95e035af7b0fb3f56805de9a0dc1fe9R120) in the time crate I don't feel very comfortable and think that the time crate probably deserves more attention by fuzzers than it has now. Or in other words, I wouldn't want to fuzz time by needling the fuzzer through the API of git_date::parse() :D. Maybe there are more direct ways to improve it.

Easy I'll see what it would involve getting the time crate integrated into oss-fuzz, and we'll see if things stabilise here as well. :)

**Comment by silvergasp** [comment]
Tracking issue in the time crate https://github.com/time-rs/time/issues/549

**Comment by silvergasp** [comment]
I haven't fully fleshed this out. But I have an idea for a new method of fuzzing. That is to generate a fuzzed input and then simultaneously pass the input into both libgit2 and gitoxide, and then cross-validate the two. Does this sound like something that will be useful?

**Comment by Byron** [comment]
Does anybody know how the [CI fuzz job](https://github.com/Byron/gitoxide/blob/d839f7c253f669f51b4227c6f5ba742925fb7f85/.github/workflows/cifuzz.yml#L15-L18) splits time across the many fuzz targets? It wasn't obvious to me when looking at the very verbose output.

Further, I noticed that more than 5 minutes are going into building the fuzzers, like it says, so the ratio of preparation to fuzzing gets worse over time apparently. All there is left when looking at effective runtimes is 3 minutes, then fuzzing will take as long as the longest job in the CI workflow. Thus fuzzing on CI could be 13 minutes long without slowing down a typical CI run, even though these 3 minutes could also be seen as buffer for when the preparation phase takes even longer.

I think the problem here is that the fuzz projects are not in the workspace, and thus can't share compilation results. Is there a reason we shouldn't put them into the workspace?

**Comment by Byron** [comment]
I think I got the answer to [one of the questions above](https://github.com/Byron/gitoxide/issues/703#issuecomment-1859643133).

It does split the [total amount of time by the amount of fuzz-targets it finds](https://github.com/google/oss-fuzz/blob/061583ebb5a96653e42feb3a97ee513eedc18078/infra/cifuzz/run_fuzzers.py#L126-L127), which is what I hoped for.

Given the recent [hang by glob here](https://github.com/Byron/gitoxide/pull/1189#issuecomment-1859948892) and the short runtime, I presume these hangs won't cause CI failures as there is not enough time for this to be flagged. 

Then only this question remains:

> I think the problem here is that the fuzz projects are not in the workspace, and thus can't share compilation results. Is there a reason we shouldn't put them into the workspace?

Thanks again for sharing your expertise.

**Comment by silvergasp** [comment]
> Given the recent hang by glob here and the short runtime, I presume these hangs won't cause CI failures as there is not enough time for this to be flagged.

This hang is kinda interesting. Do you know if globs are evaluated server side e.g. on github/gitea/gitlab. Might be worth filing an issue with git as well...

I'm having a day off today. But I'm just acknowledging your other queries, and I'll look into/respond to them in more detail tomorrow.

**Comment by Byron** [comment]
Fortunately that issue is just with `gitoxide`, and it seems I just managed to fix it 🎉. It wasn't a perfect translation after all :D.

**Comment by Byron** [comment]
By the way, `gix-glob` fuzzing would be great that way it can probably try more, and also different combinations of patterns with what's to match. After [the fixes](https://github.com/Byron/gitoxide/pull/1190) the `gix-attributes` based fuzzer was able to run for more than half an hour without issue, so superficially it appears fixed.

**Comment by silvergasp** [comment]
> Further, I noticed that more than 5 minutes are going into building the fuzzers, like it says, so the ratio of preparation to fuzzing gets worse over time apparently. All there is left when looking at effective runtimes is 3 minutes, then fuzzing will take as long as the longest job in the CI workflow. Thus fuzzing on CI could be 13 minutes long without slowing down a typical CI run, even though these 3 minutes could also be seen as buffer for when the preparation phase takes even longer.

> I think the problem here is that the fuzz projects are not in the workspace, and thus can't share compilation results. Is there a reason we shouldn't put them into the workspace?

I have been experimenting with adding fuzzer's to the workspace in a another crate recently. It is doable and it does speed things up somewhat. However there are a few caveats.
- The libfuzzer engine depends on various global instrumentation flags being used when compiling/linking. The `cargo fuzz` command takes care of this by default.
- The instrumentation flags are not available on stable and so everything has to be compiled on nightly anyway.
- You can build and run fuzz target's without instrumentation enabled on the stable channel, BUT they are next to useless as the fuzzing engine doesn't have the requisite (instrumented) coverage data it needs to do high performance fuzzing. This can lead to some weird situations where you might build/run the fuzzer but you'll get no coverage feedback and the fuzzer won't make much progress.

So as long as you are aware of those caveats, and always build the fuzzer's using `cargo +nightly fuzz build` then you can safely add them to the main workspace. If you where to do this it would;
- Speed up the fuzzer build significantly as the cache can be shared across `cargo fuzz` builds.
- Slow down the stable/normal build as it'll be building the fuzz targets as well. The fuzz target's built on the stable channel will also be effectively non-functional due to the missing instrumentation.

Perhaps a better option would be to put all fuzzer's from every crate into a single root level gix-fuzzer directory. This would share the cache between all the fuzzer builds but then also avoid mixing nightly/stable code/caching. I think you suggested this at some point (I can't find where), and at the time I said it wasn't really doable. I now think it's doable, and probably the best tradeoff in terms of build speed/correctness.

**Comment by Byron** [comment]
> Perhaps a better option would be to put all fuzzer's from every crate into a single root level gix-fuzzer directory. This would share the cache between all the fuzzer builds but then also avoid mixing nightly/stable code/caching. I think you suggested this at some point (I can't find where), and at the time I said it wasn't really doable. I now think it's doable, and probably the best tradeoff in terms of build speed/correctness.

I agree, given the caveats it seems better to find an alternative to putting fuzz-targets into the main workspace.
Putting all fuzzers together seems like it will work *and* make invoking fuzzers for various crates more convenient, it's all in one spot, without so much duplication maybe. On the other hand, the initial build would then take longer as there is just one Cargo manifest, which will build all dependencies, once. For CI, this is not a problem though, and locally it's probably fine as well.

As an immediate solution, maybe one could experiment with setting the `CARGO_BUILD_TARGET_DIR` environment variable for all steps to share the build output that way. Maybe this immediately speeds up build times without changing anything else?

If you think this could work, I'd definitely appreciate a quick PR to try that out, otherwise I'd go with the top-level fuzz project (which wouldn't be required if the environment variable already does the trick.

**Comment by silvergasp** [comment]
> As an immediate solution, maybe one could experiment with setting the CARGO_BUILD_TARGET_DIR environment variable for all steps to share the build output that way. Maybe this immediately speeds up build times without changing anything else?

Easy, I'm running an experiment now to see how that pans out https://github.com/google/oss-fuzz/pull/11385

**Comment by silvergasp** [comment]
Results from my local testing with an 8 core laptop. Note that this isn't a perfect benchmark as it includes downloading crates, on a fairly unreliable internet connection.

_OSS fuzz command_:
`time python infra/helper.py build_fuzzers --sanitizer address gitoxide`

## Before
```
real	4m37.855s
user	0m0.650s
sys	0m0.456s
```

## After
```
real	4m3.367s
user	0m0.620s
sys	0m0.406s
```

## Delta
real -34.9s

**Comment by silvergasp** [comment]
So it's a bit of a speedup but not as much as I thought it would be. It might be more pronounced on github actions as they use less cores and have a faster/more-reliable internet connection.

**Comment by Byron** [comment]
Yes, indeed much less than I thought, but then again it's probably worth seeing what CI does. If in doubt, it's an easy improvement to have and 30s count for something, for sure.

**Comment by silvergasp** [comment]
Aight I'll mark that PR as read for review.

**Comment by Byron** [comment]
Thanks! Do you think it's worth trying the same for fuzz-CI job here? If so, feel free to put it into [this PR](https://github.com/Byron/gitoxide/pull/1188) if that reduces overhead.

**Comment by silvergasp** [comment]
> Thanks! Do you think it's worth trying the same for fuzz-CI job here? If so, feel free to put it into [this PR](https://github.com/Byron/gitoxide/pull/1188) if that reduces overhead.

The CI Fuzzer uses the same [config from oss fuzz](https://google.github.io/oss-fuzz/getting-started/continuous-integration/) so, once the upstream PR is accepted it'll also be reflected here. The CIFuzzer is effectively a stripped down version of oss-fuzz/clusterfuzz.

**Comment by silvergasp** [comment]
I've been looking at fuzzing more generally, and it's looking like we are almost through the high value low hanging fruit. I've been mostly using the [coverage feedback](https://storage.googleapis.com/oss-fuzz-coverage/gitoxide/reports/20231220/linux/src/gitoxide/report.html) to identify API's that when fuzzed will have a high impact (in lines of code). The last big chunk that is not covered at all is `gix-ref` but I'm pretty certain with the recent changes these should start being covered by tommorow.

The remaining high value API's are looking like they might need a little bit of refactoring to make them easier to fuzz test (e.g. https://github.com/Byron/gitoxide/pull/1188#discussion_r1431739805). The main difficulties come from types that can only be constructed from files/pathbufs. I would like to discuss a few approaches to refactoring before I go make broad sweeping changes.
1. Where possible replace api's that construct from `PathBuf`, `Path`, `File` with inputs that implement `std::io::{Read, Write, Seek}`  traits. In some cases you'll still need that actual path as well. It should be possible to keep this backwards compatible, with the original API.
2. Another alternative which is what we did with https://github.com/Byron/gitoxide/pull/1188#discussion_r1431739805 is to just create an API that takes `&[u8]` or similar.
3. Replace all `PathBuf` inputs with `File` objects this would mean I'd be able to use the `tempfile` crate to dump fuzzed data into the the `/tmp` directory.

I think any additional API's could be guarded behind `#cfg(fuzzing)` if you aren't keen on exposing these constructors by default.

**Comment by silvergasp** [comment]
On a different note, I'm genuinely surprised that given the number of fuzzer's there is still fairly [low coverage](https://storage.googleapis.com/oss-fuzz-coverage/gitoxide/reports/20231220/linux/src/gitoxide/report.html), we are currently looking at around ~15% coverage. For comparison I've previously written a single fuzzer for a large project and achieved ~65% coverage. At the moment we are at around 15 fuzzer's total.

Are there any functions that you think could be easily fuzzable that would touch a broad spread of other API's. i.e. a fuzzable target that would sit at the top of a large call-tree. I'm hoping to be a little more efficient in terms of my efforts and the actual impact.

**Comment by Byron** [comment]
> I've been looking at fuzzing more generally, and it's looking like we are almost through the high value low hanging fruit. I've been mostly using the [coverage feedback](https://storage.googleapis.com/oss-fuzz-coverage/gitoxide/reports/20231220/linux/src/gitoxide/report.html) to identify API's that when fuzzed will have a high impact (in lines of code). The last big chunk that is not covered at all is `gix-ref` but I'm pretty certain with the recent changes these should start being covered by tommorow.
> 
> The remaining high value API's are looking like they might need a little bit of refactoring to make them easier to fuzz test (e.g. [#1188 (comment)](https://github.com/Byron/gitoxide/pull/1188#discussion_r1431739805)). The main difficulties come from types that can only be constructed from files/pathbufs. I would like to discuss a few approaches to refactoring before I go make broad sweeping changes.
> 
> 1. Where possible replace api's that construct from `PathBuf`, `Path`, `File` with inputs that implement `std::io::{Read, Write, Seek}`  traits. In some cases you'll still need that actual path as well. It should be possible to keep this backwards compatible, with the original API.
> 2. Another alternative which is what we did with [gix-ref: Add fuzz-harness's #1188 (comment)](https://github.com/Byron/gitoxide/pull/1188#discussion_r1431739805) is to just create an API that takes `&[u8]` or similar.
> 3. Replace all `PathBuf` inputs with `File` objects this would mean I'd be able to use the `tempfile` crate to dump fuzzed data into the the `/tmp` directory.

That's a good idea! I think the approach taken should depend on what the consumer of the `PathBuf` is doing with it. If they end up storing everything in memory, the alternative constructor could just provide what they want. Thus I wouldn't generalize an approach but always tailor it so that it fits the type at hand.

> 
> I think any additional API's could be guarded behind `#cfg(fuzzing)` if you aren't keen on exposing these constructors by default.

That's good to have, especially since I wasn't aware that the fuzzer compiles with this `cfg` available. By default, I will be fine with adding these extra functions to the API surface as long as they are in a plumbing crate - what's useful for the fuzzer can be useful to direct users of that crate as well.

**Comment by Byron** [comment]
> On a different note, I'm genuinely surprised that given the number of fuzzer's there is still fairly [low coverage](https://storage.googleapis.com/oss-fuzz-coverage/gitoxide/reports/20231220/linux/src/gitoxide/report.html), we are currently looking at around ~15% coverage. For comparison I've previously written a single fuzzer for a large project and achieved ~65% coverage. At the moment we are at around 15 fuzzer's total.

I'd think these percentages are deceptive unless the underlying line count is also the same. But even if they were, I'd be surprised if this was possible here as the API surface is huge. If one day there would be a `git coffee` command I wouldn't even be surprised, it can do pretty much everything 😅.

> Are there any functions that you think could be easily fuzzable that would touch a broad spread of other API's. i.e. a fuzzable target that would sit at the top of a large call-tree. I'm hoping to be a little more efficient in terms of my efforts and the actual impact.

It's certainly wise to try to go for coverage, but to me that would be a secondary factor. The primary one would be related to covering all code that gets in touch with untrusted input. That starts from the initial clone, and continues to code that runs when interacting with repositories that might have been altered by an adversary.

This would probably mean an order like this for the initial clone:

* `gix-packetline` - the way git encodes data on the wire
* `gix-protocol` - parsing of types that are exchanged on the wire, for example for fetching
* `gix-pack` - parse and handle a pack transmitted over the wire
* `gix-index` - it has to create an index from the root tree
* `gix-worktree-state` - perform the initial checkout
* `gix-ref` for writing out the first set of references

Then for accessing the data in the repository, there is:

* `gix-config` - all kinds of git configuration files
* `gix-object`
* `gix-quote`
* `gix-glob` - used by `gix-ignore` and `gix-attributes`
* `gix-odb` - note there is also the 'alternates' file format
* `gix-ref` - packed refs and normal refs
* `gix-commitgraph` - some repos have it to accelerate commit access

And of course, there is much more that gets triggered when the user performs some operations, but abuse seems less likely there.

I hope that helps.

**Comment by silvergasp** [comment]
Cool thanks for the run down, that gives me a little more direction

**Comment by silvergasp** [comment]
I was hoping to get your thoughts on using some sort of virtual filesystem with gix-* this would allow for easier testing/fuzzing as you can fake out an in memory virtual filesystem, it'd also allow for things like file-system sand-boxing e.g. optionally constraining gitoxide to a subset of the filesystem. There is already an implementation of something like this here https://github.com/manuel-woelker/rust-vfs.

**Comment by Byron** [comment]
Thanks for sharing! 

I took a look and didn't immediately see how `rust-vfs` could be integrated transparently, which probably would be a optimal. Further, there are plenty of uses of `mmap2`, which is unlikely to play well with `rust-vfs`. Lastly, the `VfsPath` for the best of all reasons uses UTF-8 encoded bytes for storage, which would be a problem for Git on Linux where there is no imposed encoded on file paths.

Maybe `rust-vfs` can be useful in a more limited capacity though, I just don't see it yet :/.

**Comment by silvergasp** [comment]
> I took a look and didn't immediately see how `rust-vfs` could be integrated transparently, which probably would be a optimal. Further, there are plenty of uses of `mmap2`, which is unlikely to play well with `rust-vfs`. Lastly, the `VfsPath` for the best of all reasons uses UTF-8 encoded bytes for storage, which would be a problem for Git on Linux where there is no imposed encoded on file paths.

Cool thanks for taking a look. I was mostly interested coming from the perspective of fuzzing. But it would be a non-trivial amount of work to migrate to a VFS, and it probably isn't worthwhile unless there was significant benefits in other areas. Which doesn't sound like it's the case.

**Comment by silvergasp** [comment]
Another question. Do you have a specific contract around how your error types are displayed. i.e.
```rust
let error_message = format!("{error}");
// Can I say that error_message is all ascii?
assert!(error_message.is_ascii());
// Can I say that error_message is free of control codes (e.g. clear line, bold, italic, flashing, color codes etc.)
assert!(no_control_codes(error_message));
```
I'm mostly trying to work out how I can fuzz the error handling code. At the moment these types are largely being discarded.

**Comment by Byron** [comment]
> I'm mostly trying to work out how I can fuzz the error handling code. At the moment these types are largely being discarded.

I'd recommend to continue doing so. The error handling code itself is naturally run by the fuzzer, more so than other code probably, and the error types are all generated. Thus I wouldn't use any CPU time to look at them.

There are interesting 'error' scenarios which is spilling of sensitive information using errors or traces, but for finding these I'd think the fuzzer isn't the tool of choice. When thinking about this, `gix-url` is probably the crate to review and to assure the display of a URl doesn't show passwords.

On another note, the `thiserror` powered error types are an issue for the stability of any crate, as adding any now variant to the enum will break the API. Thus I hope to replace it with something else that makes errors more generic, but still introspectable. That then would certainly be [a separate crate](https://github.com/Byron/therror) to fuzz, even though work on it stalled and probably has to be reinvigorated when it's time to stabilize `gix*`.

**Comment by silvergasp** [comment]
> I'd recommend to continue doing so. The error handling code itself is naturally run by the fuzzer, more so than other code probably, and the error types are all generated. Thus I wouldn't use any CPU time to look at them.

Cool good to know.

> There are interesting 'error' scenarios which is spilling of sensitive information using errors or traces, but for finding these I'd think the fuzzer isn't the tool of choice. When thinking about this, `gix-url` is probably the crate to review and to assure the display of a URl doesn't show passwords.

I think this sounds pretty interesting. I've been experimenting with fuzzing for CTFs. In this case, the "flag" would be a password. I'll just be a little more cautious with submitting fuzzing PRs as that would constitute a real vulnerability as apposed to the the relatively harmless bugs I've found so fair.

> On another note, the `thiserror` powered error types are an issue for the stability of any crate, as adding any now variant to the enum will break the API. Thus I hope to replace it with something else that makes errors more generic, but still introspectable. That then would certainly be [a separate crate](https://github.com/Byron/therror) to fuzz, even though work on it stalled and probably has to be reinvigorated when it's time to stabilize `gix*`.

I'll keep a look out, feel free to ping me with a message if you'd like to prioritise X crate when stabilising a crate.

**Comment by Byron** [comment]
This is actually done!
